### ISSUE TITLE
Upgrading from 0.1.9 to 0.2.0 silently breaks policy gradient code?

### ISSUE THREAD

**Issue by p-morais** [pytorch/pytorch#2598]
I was implementing Proximal Policy Optimization when I noticed that my Pytorch version was outdated, so I updated. To my surprise, the code I was running which worked fine in 0.1.9 was completely broken in 0.2.0 (i.e. it ran without any errors or warnings, but the policy updates suddenly weren't improving the surrogate loss or expected reward at all). 

Here is a minimal example that reproduces the problem:

```python
import torch
import torch.nn as nn
from torch.autograd import Variable
from torch.optim import Adam
import torch.nn.functional as F


import numpy as np
import gym

class GaussianMLP(nn.Module):
    def __init__(self, obs_dim, action_dim):

        super(GaussianMLP, self).__init__()

        self.hidden_layer = nn.Linear(obs_dim, 8)

        self.means = nn.Linear(8, action_dim)

        self.log_stds = nn.Parameter(
            torch.ones(1, action_dim)
        )

    def forward(self, x):
        x = F.tanh(self.hidden_layer(x))

        means = self.means(x)

        log_stds = self.log_stds.expand_as(means)

        stds = torch.exp(log_stds)

        return means, log_stds, stds

    def log_likelihood(self, x, means, log_stds, stds):
        var = stds.pow(2)

        log_density = -(x - means).pow(2) / (
            2 * var) - 0.5 * np.log(2 * np.pi) - log_stds

        return log_density.sum(1)

    def get_action(self, means, stds):
        action = torch.normal(means, stds)
        return action.detach()


def train(n_itr,
          n_trj,
          max_trj_len,
          env,
          policy,
          optimizer,
          discount=0.99):

    def rollout(env, policy, max_trj_len):
        """Collect a single rollout."""
        rewards = []
        observations = []
        actions = []
        returns = []

        obs = env.reset().ravel()[None, :] # reshape because of OpenAI gym inconsistencies
        for _ in range(max_trj_len):
            obs_var = Variable(torch.Tensor(obs))

            means, log_stds, stds = policy(obs_var)
            action = policy.get_action(means, stds)

            next_obs, reward, done, _ = env.step(action.data.numpy().ravel())

            observations.append(obs_var)
            actions.append(action)
            rewards.append(Variable(torch.Tensor([reward])))

            obs = next_obs.ravel()[None, :]

            if done:
                break

        R = Variable(torch.zeros(1, 1))
        for i in reversed(range(len(rewards))):
            R = discount * R + rewards[i]
            returns.append(R)

        return dict(
            rewards=torch.stack(rewards),
            returns=torch.cat(returns[::-1]),
            observations=torch.cat(observations),
            actions=torch.cat(actions)
        )

    for _ in range(n_itr):
        paths = [rollout(env, policy, max_trj_len) for _ in range(n_trj)]

        observations = torch.cat([p["observations"] for p in paths])
        actions = torch.cat([p["actions"] for p in paths])
        returns = torch.cat([p["returns"] for p in paths])

        means, log_stds, stds = policy(observations)

        logprobs = policy.log_likelihood(actions, means, log_stds, stds)

        policy_loss = -(logprobs * returns).mean()

        optimizer.zero_grad()
        policy_loss.backward()
        optimizer.step()

        print(
            'Average Return: %f' %
            np.mean(([p["rewards"].sum().data.numpy() for p in paths]))
        )



if __name__ == "__main__":
    env = gym.make("Hopper-v1")

    obs_dim = env.observation_space.shape[0]
    action_dim = env.action_space.shape[0]

    policy = GaussianMLP(obs_dim, action_dim)

    train(
        n_itr=100,
        n_trj=200,
        max_trj_len=100,
        env=env,
        policy=policy,
        optimizer=Adam(policy.parameters(), lr=0.01)
    )
```
At least on my machine running the above code in 0.1.9 yields stable reward improvement, whereas in 0.2.0 the rewards stays the same, with minor fluctuations (indicating the gradient is wrong?). What is going on?

**Comment by soumith** [comment]
@p-morais 0.2.0 has some breaking changes that can be detected if you insert a piece of code to the top of your scripts. Warnings will be generated in such cases. We've put this in the 0.2.0 release notes, in the mailing list etc.
The breaking changes are due to introduction of broadcasting, and changing the way dimensions are collapsed during reductions (and making them consistent with numpy)

I'd urge you to check if your code is affected by these changes. Instructions are in the release notes here:
https://github.com/pytorch/pytorch/releases/tag/v0.2.0
See the "Important Breakages and Workarounds" section.

let us know if that does not fix your issue, and we can investigate further.

**Comment by p-morais** [comment]
Ah, I missed the silent breakages notice, my bad. After following the instructions and fixing some broadcasting changes it appears to work the same in both versions.

I'll close this issue now. Thanks!
